overlord/snapstate/target.go

```diff
@@ -266,37 +285,96 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
 			snapst = &SnapState{}
 		}
 
-		// TODO: extract components from resources
-
 		// TODO: is it safe to pull the channel from here? i'm not sure what
 		// this will actually look like as a response from the real store
 		channel := r.RedirectChannel
 		if r.RedirectChannel == "" {
 			channel = sn.RevOpts.Channel
 		}
 
+		comps, err := requestedComponentsFromActionResult(sn, r)
+		if err != nil {
+			return nil, fmt.Errorf("cannot extract components from snap resources: %w", err)
+		}
+
 		installs = append(installs, target{
-			setup: &SnapSetup{
+			setup: SnapSetup{
 				DownloadInfo: &r.DownloadInfo,
 				Channel:      channel,
 				CohortKey:    sn.RevOpts.CohortKey,
 			},
 			info:       r.Info,
 			snapst:     *snapst,
-			components: nil,
+			components: comps,
 		})
 	}
 
 	return installs, err
 }
 
+func requestedComponentsFromActionResult(sn StoreSnap, sar store.SnapActionResult) ([]componentTarget, error) {
+	mapping := make(map[string]store.SnapResourceResult, len(sar.Resources))
+	for _, res := range sar.Resources {
+		mapping[res.Name] = res
+	}
+
+	installables := make([]componentTarget, 0, len(sn.Components))
+	for _, comp := range sn.Components {
+		res, ok := mapping[comp]
+		if !ok {
+			return nil, fmt.Errorf("cannot find component %q in snap resources", comp)
+		}
+
+		installable, err := componentFromResource(comp, res, sar.Info)
+		if err != nil {
+			return nil, err
+		}
+
+		installables = append(installables, installable)
+	}
+	return installables, nil
+}
+
+func componentFromResource(name string, sar store.SnapResourceResult, info *snap.Info) (componentTarget, error) {
+	comp, ok := info.Components[name]
+	if !ok {
+		return componentTarget{}, fmt.Errorf("%q is not a component for snap %q", name, info.SnapName())
+	}
+
+	if typ := fmt.Sprintf("component/%s", comp.Type); typ != sar.Type {
+		return componentTarget{}, fmt.Errorf("inconsistent component type (%q in snap, %q in component)", typ, sar.Type)
+	}
+
+	compName := naming.NewComponentRef(info.SnapName(), name)
+
+	csi := snap.ComponentSideInfo{
+		Component: compName,
+		Revision:  snap.R(sar.Revision),
+	}
+
+	return componentTarget{
+		setup: ComponentSetup{
+			DownloadInfo: &sar.DownloadInfo,
+			CompSideInfo: &csi,
+			CompType:     comp.Type,
+		},
+		info: &snap.ComponentInfo{
+			Component:         compName,
+			Type:              comp.Type,
+			Version:           sar.Version,
+			ComponentSideInfo: csi,
+		},
+	}, nil
+}
+
 func installActionForStoreTarget(t StoreSnap, opts Options, enforcedSets func() (*snapasserts.ValidationSets, error)) (*store.SnapAction, error) {
 	action := &store.SnapAction{
 		Action:       "install",
 		InstanceName: t.InstanceName,
 		Channel:      t.RevOpts.Channel,
 		Revision:     t.RevOpts.Revision,
 		CohortKey:    t.RevOpts.CohortKey,
+		Resources:    t.Components,
```

Comment:
what's the reason we need to pass this along? for metrics?

Reply:
It is ultimately used to indicate in the store package that we should request resource info from the store. It could (and maybe should?) be a boolean value. It is a bit weird,  since the option is a per-request toggle (just a query param), but right here I have it expressed as a per-action concept. But the `StoreService.SnapAction` API doesn't have any sort of flags right now.